src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java

```diff
@@ -93,61 +103,133 @@ public KNNQueryBuilder(String fieldName, float[] vector) {
     }
 
     /**
-     * Builder method for k
-     *
-     * @param k K nearest neighbours for the given vector
+     * lombok SuperBuilder annotation requires a builder annotation on parent class to work well
+     * {@link AbstractQueryBuilder#boost()} and {@link AbstractQueryBuilder#queryName()} both need to be called
+     * A custom builder helps with the calls to the parent class, simultaneously addressing the problem of telescoping
+     * constructors in this class.
      */
-    public KNNQueryBuilder k(Integer k) {
-        if (k == null) {
-            throw new IllegalArgumentException(String.format("[%s] requires k to be set", NAME));
+    public static class Builder {
```

Comment:
Good Improvement to use Builder ..but we can avoid writing whole stuff by our self with little improvement 


``
```
@AllArgsConstructor
@Getter
@Setter
@Builder
public class KNNQueryBuilder extends AbstractQueryBuilder<KNNQueryBuilder> {
    // Existing fields remain unchanged
    @NonNull
    private final String fieldName;
    @NonNull
    private final float[] vector;
    @Builder.Default
    private int k = 0;
    private Integer efSearch;
    private Float maxDistance;
    private Float minScore;
    private QueryBuilder filter;
    @Builder.Default
    private boolean ignoreUnmapped = false;
    private String queryName;
    private float boost = DEFAULT_BOOST;

    // Custom validation method
    private void validate() {
        if (k < 0 || k > K_MAX) {
            throw new IllegalArgumentException(String.format("[%s] requires k to be in the range [0, %d]", NAME, K_MAX));
        }

        // Add more validation rules as needed
    }

    // Builder method for k with custom validation
    public KNNQueryBuilder k(int k) {
        this.k = k;
        validate();
        return this;
    }

    // Other builder methods remain unchanged
}
```

Reply:
1. This is not going to work as mentioned in the comment above the `Builder` class definition. There is a parent class involved there which holds certain setter methods which won't seamlessly work with lombok. I wouldn't want to make changes in opensearch repo as there might be many queries using it.

Considering telescoping constructors, its worth maintaining a builder than multiple constructors which is more error prone and unreadable.

2.
```
public KNNQueryBuilder k(int k) { this.k = k; validate(); return this; }
```
This opens up a way to modify k (or any other request parameter), which I don't think is a good idea considering this object goes to parts of code that are not owned by us. This is an object which is formed after parsing json request. With buillder class and no setters I am making it immutable in a way.

3. Individual validations are avoided as suggested because there are validations which involve a combination of variables.